base_layer/core/tests/block_validation.rs

```diff
@@ -52,3 +66,206 @@ fn test_genesis_block() {
+    assert_eq!(db.add_block(block.clone()), Ok(BlockAddResult::Ok));
+    blocks.push(block);
+    // Block 2
+
+    let schema = vec![txn_schema!(from: vec![outputs[1][0].clone()], to: vec![3 * T, 1 * T])];
+    let coinbase_amount = 326 * uT + rules.emission_schedule().block_reward(2);
+    let mut block =
+        calculate_new_block(&mut db, &factories, &mut blocks, &mut outputs, schema, coinbase_amount).unwrap();
+    block.header.timestamp = EpochTime::now().checked_sub(EpochTime::from(3_000)).unwrap();
+    find_header_with_achieved_difficulty(&mut block.header, Difficulty::from(100));
+
+    block.header.timestamp = EpochTime::now().checked_sub(EpochTime::from(4_000)).unwrap();
+    assert_eq!(db.add_block(block.clone()), Ok(BlockAddResult::Ok));
+    blocks.push(block);
+    // Block 3
+
+    let schema = vec![
+        txn_schema!(from: vec![outputs[2][0].clone()], to: vec![2 * T, 500_000 * uT]),
+        txn_schema!(from: vec![outputs[1][1].clone()], to: vec![500_000 * uT]),
+    ];
+    let coinbase_amount = 326 * uT + 226 * uT + rules.emission_schedule().block_reward(3);
+    let mut block =
+        calculate_new_block(&mut db, &factories, &mut blocks, &mut outputs, schema, coinbase_amount).unwrap();
+
+    dbg!(&block.body.kernels());
+
+    block.header.timestamp = EpochTime::now().checked_sub(EpochTime::from(2_000)).unwrap();
+    find_header_with_achieved_difficulty(&mut block.header, Difficulty::from(200));
+    assert_eq!(db.add_block(block.clone()), Ok(BlockAddResult::Ok));
+}
+
+#[test]
+fn test_invalid_coinbase() {
+    let factories = CryptoFactories::default();
+    let rules = ConsensusManager::default();
+    let backend = MemoryDatabase::<HashDigest>::default();
+    let mut db = BlockchainDatabase::new(backend).unwrap();
+    let validators_true = Validators::new(
+        FullConsensusValidator::new(rules.clone(), factories.clone(), db.clone()),
+        StatelessValidator::new(),
+        MockValidator::new(true),
+        MockValidator::new(true),
+        MockValidator::new(true),
+    );
+    let validators_false = Validators::new(
+        MockValidator::new(true),
+        MockValidator::new(true),
+        MockValidator::new(true),
+        MockValidator::new(true),
+        MockValidator::new(true),
+    );
+    db.set_validators(validators_false);
+    let diff_adj_manager = DiffAdjManager::new(db.clone()).unwrap();
+    rules.set_diff_manager(diff_adj_manager).unwrap();
+
+    let (block0, output) = create_genesis_block_with_utxos(&db, &factories, &[10 * T]);
+    db.add_block(block0.clone()).unwrap();
+    let mut blocks = vec![block0];
+    let mut outputs = vec![output];
+    db.set_validators(validators_true);
+    let schema = vec![txn_schema!(from: vec![outputs[0][1].clone()], to: vec![6 * T, 3 * T])];
+    // We have no coinbase, so this should fail
+    assert!(generate_new_block(&mut db, &mut blocks, &mut outputs, schema).is_err());
+}
```

Comment:
I suggest updating `sample_blockchains::create_new_blockchain` to remove all this boilerplate